// ngx-typewriter/esm2015/lib/ngx-typewriter.service.js
import { Injectable } from '@angular/core';
import { TPW } from "../models/TPW.model";
import * as i0 from "@angular/core";
export class NgxTypewriterService {
    constructor() {
        this.list = {};
    }
    getTPW(selector, options) {
        if (!this.list.hasOwnProperty(selector) && options) {
            this.list[selector] = new TPW(options);
        }
        return this.list[selector] || null;
    }
}
NgxTypewriterService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NgxTypewriterService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
