// ngx-typewriter/esm2015/lib/ngx-typewriter.module.js
import { NgModule } from '@angular/core';
import { NgxTypewriterComponent } from './ngx-typewriter.component';
import * as i0 from "@angular/core";
export class NgxTypewriterModule {
}
NgxTypewriterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxTypewriterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterModule, declarations: [NgxTypewriterComponent], exports: [NgxTypewriterComponent] });
NgxTypewriterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxTypewriterComponent,
                    ],
                    imports: [],
                    exports: [
                        NgxTypewriterComponent
                    ]
                }]
        }] });
