// ngx-typewriter/esm2015/lib/ngx-typewriter.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-typewriter.service";
import * as i2 from "./style.service";
export class NgxTypewriterComponent {
    constructor(tpWriterService, styleService) {
        this.tpWriterService = tpWriterService;
        this.styleService = styleService;
        this.selector = 'tpw';
        this.marker = '|';
        NgxTypewriterComponent.styleService = this.styleService;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.tpw = this.tpWriterService.getTPW(this.selector, this.options);
            for (const className of Object.keys(this.tpw.classList)) {
                this.styleService.setStyles(`.${className}`, this.tpw.classList[className]);
            }
            yield this.tpw.runDisplay();
        });
    }
}
NgxTypewriterComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterComponent, deps: [{ token: i1.NgxTypewriterService }, { token: i2.StyleService }], target: i0.ɵɵFactoryTarget.Component });
NgxTypewriterComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.5", type: NgxTypewriterComponent, selector: "tp-writer", inputs: { options: "options", selector: "selector", marker: "marker" }, ngImport: i0, template: `
    <span [style.color]="tpw.color ? tpw.color : null" [innerHTML]="tpw.displayText"></span>
    <span class="marker">{{marker}}</span>
  `, isInline: true, styles: [".marker { animation: 0.5s ease 0s infinite normal none running blink; } @keyframes blink {from {opacity: 0;} to {opacity: 1;}}"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: NgxTypewriterComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tp-writer',
                    template: `
    <span [style.color]="tpw.color ? tpw.color : null" [innerHTML]="tpw.displayText"></span>
    <span class="marker">{{marker}}</span>
  `,
                    styles: ['.marker { animation: 0.5s ease 0s infinite normal none running blink; } @keyframes blink {from {opacity: 0;} to {opacity: 1;}}']
                }]
        }], ctorParameters: function () { return [{ type: i1.NgxTypewriterService }, { type: i2.StyleService }]; }, propDecorators: { options: [{
                type: Input
            }], selector: [{
                type: Input
            }], marker: [{
                type: Input
            }] } });
