// ngx-typewriter/esm2015/lib/style.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class StyleService {
    constructor() { }
    get headElement() {
        if (!this._headElement && document.getElementsByTagName("head").length > 0) {
            this._headElement = document.getElementsByTagName("head")[0];
        }
        return this._headElement;
    }
    get cssStyleSheet() {
        if (!this._cssStyleSheet) {
            if (!document.styleSheets || this.headElement === null)
                return null;
            // Get the first style sheet that is enabled and mediaText is empty or screen.
            this._cssStyleSheet = Array.from(document.styleSheets).find(s => !s.disabled && (s.media.mediaText == "" || s.media.mediaText.indexOf("screen") !== -1));
            // If the style sheet doesn't exist yet, then create it.
            if (!this._cssStyleSheet)
                this._cssStyleSheet = this.createCssStyleSheet();
        }
        return this._cssStyleSheet;
    }
    setStyles(selectorText, styles) {
        let rule = this.getStyleRule(selectorText);
        if (!rule)
            return;
        Object.keys(styles).forEach(styleName => {
            // @ts-ignore
            rule.style[styleName] = styles[styleName];
        });
    }
    createCssStyleSheet() {
        // Create the style sheet element.
        let styleSheetElement = document.createElement("style");
        styleSheetElement.type = "text/css";
        // Append the style sheet element to the head.
        this.headElement.appendChild(styleSheetElement);
        return styleSheetElement.sheet;
    }
    getStyleRule(selectorText) {
        if (!this.cssStyleSheet) { // @ts-ignore
            return;
        }
        let rules = this.cssStyleSheet.cssRules.length > 0 || this.cssStyleSheet.rules.length == 0 ? this.cssStyleSheet.cssRules : this.cssStyleSheet.rules;
        let rule = Array.from(rules).find(r => r instanceof CSSStyleRule && r.selectorText.toLowerCase() == selectorText.toLowerCase());
        // If the selector rule does not exist, create it.
        if (!rule) {
            let ruleIndex = this.cssStyleSheet.insertRule(selectorText + "{ }", rules.length);
            rule = rules[ruleIndex];
        }
        return rule;
    }
}
StyleService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: StyleService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
StyleService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: StyleService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: StyleService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
